### PR TITLE
Add serde support to Mime type

### PR DESCRIPTION
**Pull request by novacrazy**
This PR adds simple but functional `Serialize` and `Deserialize` implementations for the `Mime` structure using the `as_ref` and `from_str` methods.

I also changed the `.travis.yml` configuration slightly to build/test all features, and fixed an "unused return value" error in the tests.

### CODE REVIEW

**Comment by novacrazy** [comment]
Is this project still maintained?

**Comment by seanmonstar** [comment]
Sorry, yes it's maintained, this slipped through my inbox.

I've been a little apprehensive in the past about adding Serde support directly in crates, as I don't want breaking changes in Serde to mean breaking changes in these libs also. That combined with serde's ability to use `serialize_with` and `deserialize_with` has made me so far put this off.

**Comment by novacrazy** [comment]
Okay. I promise this is in no way passive-aggressive, but I went ahead and created a new crate specifically to host shim functions for other crates, with Mime being the first because I actually need it now.

See it [on crates.io](https://crates.io/crates/serde_shims)

If that sounds cool with you, I can promote it on reddit/IRC and get more shims incorporated.

**Comment by xfix** [comment]
What about https://rust-lang-nursery.github.io/api-guidelines/interoperability.html#data-structures-implement-serdes-serialize-deserialize-c-serde?

File: src/serde_impl.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a valid MIME type")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<Mime, E>
+            where
+                E: de::Error,
+            {
+                Mime::from_str(value).or_else(|e| Err(E::custom(format!("{}", e))))
```
Comment (xfix) [inline src/serde_impl.rs]: Cannot this be simply `Mime::from_str(value).or_else(|e| Err(E::custom(e)))`, as in, with removed `format!` call?

File: Cargo.toml [hunk 2]
```diff
@@ -14,3 +14,10 @@ publish = false # breaking changes from 0.3.x
 [dependencies]
 unicase = "2.0"
 quoted-string = "0.2.2"
+
+[dependencies.serde]
+optional = true
+version = "1.0"
+
+[dev-dependencies]
+serde_json = "1.0.17"
```
Comment (xfix) [inline Cargo.toml]: You could add a newline here, to avoid "No newline at end of file" warning.

File: src/lib.rs [hunk 3]
```diff
@@ -32,6 +32,9 @@
 extern crate unicase;
 extern crate quoted_string;
 
+#[cfg(feature = "serde")]
+extern crate serde;
```
Comment (xfix) [inline src/lib.rs]: I wonder if it's possible to use `extern crate serde` directly in `serde_impl` mod.

File: src/lib.rs [hunk 4]
```diff
@@ -555,7 +561,7 @@ macro_rules! mimes {
 
         #[test]
         fn test_mimes_macro_consts() {
-            [
+            let _ = [
```
Comment (xfix) [inline src/lib.rs]: Why did you put this assignment here, just curious?

File: src/lib.rs [hunk 5]
```diff
@@ -891,5 +897,19 @@ mod tests {
         let mime2 = Mime::from_str(r#"text/x-custom; aBc=a"#).unwrap();
         assert_eq!(mime1, mime2);
     }
+
+    #[cfg(feature = "serde")]
+    #[test]
+    fn serde_tests() {
+        extern crate serde_json;
```
Comment (xfix) [inline src/lib.rs]: You may want to use `serde_test` crate for testing purposes.

File: src/lib.rs [hunk 6]
```diff
@@ -891,5 +897,19 @@ mod tests {
         let mime2 = Mime::from_str(r#"text/x-custom; aBc=a"#).unwrap();
         assert_eq!(mime1, mime2);
     }
+
+    #[cfg(feature = "serde")]
+    #[test]
+    fn serde_tests() {
```
Comment (xfix) [inline src/lib.rs]: Probably should be called `test_serde`, as this seems to be naming convention in this crate.

File: src/lib.rs (hunk 4 above)
Comment (novacrazy) [reply src/lib.rs]: It fixed an "unused return value" error in the tests. That's why some of the other pull requests are failing CI, too.

**Comment by erickt** [comment]
Hi @seanmonstar! Serde is following semver rules, and so now that serde is 1.x, we won't break backwards compatibility (unless we're forced to due to a soundness bug in the compiler). You shouldn't need to worry about us breaking you.

That said, getting your users to use `serialize_with` and `deserialize_with` is also a fine choice.

**Comment by seanmonstar** [comment]
@novacrazy:

> Okay. I promise this is in no way passive-aggressive, but I went ahead and created a new crate specifically to host shim functions for other crates, with Mime being the first because I actually need it now.

Not a problem at all, that's exactly what I mentioned in my comment :D

@xfix:

> What about https://rust-lang-nursery.github.io/api-guidelines/interoperability.html#data-structures-implement-serdes-serialize-deserialize-c-serde? Serde support is part of official Rust API guidelines.

That's OK, I don't agree with that one and a couple other guidelines in there ;).

@erickt: Oh I understand, I definitely have no problem using Serde in applications. I know it follows semver, but that doesn't mean breaking changes aren't a thing. I mean that if eventually Serde 2.0 is needed, libraries showing `serde` as a public dependency need to *also* release a new "breaking change" version, or else not support the new stuff.

My opinion as a maintainer is that it's easier for stability to use glue crates.

**Comment by erickt** [comment]
@seanmonstar definitely. We'd need a *really* good reason to do a serde 2.x though, since at this point it'd cause another libcpocalypse-style event. Glue crates is a fine way to avoid this.

**Comment by ebkalderon** [comment]
Perhaps if you're worried about upstream serde changes causing breakage, you could make `serde` an opt-in dependency gated by a feature and add the following conditional attribute to the relevant structs:

```rust
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
```

This way, only dependent crates which explicitly opt into serde support for mime will pull it in as a dependency. If serde happens to break, crates not explicitly opting into it will remain unaffected. Would this potentially be a good compromise?

**Comment by novacrazy** [comment]
@ebkalderon This PR already does that.

**Comment by novacrazy** [comment]
So I guess this isn't going to happen?

**Comment by nifker** [comment]
This would be very sad if this crate won't support it as I then would need to impl Deserialize for a struct myself.

**Comment by novacrazy** [comment]
@nifker I've already implemented it in the [`serde_shims`](https://crates.io/crates/serde_shims) crate, or more specifically the [`mime_serde_shim`](https://crates.io/crates/mime_serde_shim) crate.

**Comment by nifker** [comment]
@novacrazy this sounds like a good solution for the meantime ^^

**Comment by nifker** [comment]
@novacrazy I cant access mime through your crate as it seems to be 'private'.

**Comment by novacrazy** [comment]
@nifker Odd. Can you open an issue on https://github.com/novacrazy/serde_shims so we don't clutter up this one?

**Pull request closed without merging**